core/shared/src/main/scala/scalaz/zio/Async.scala

```diff
@@ -1,60 +1,29 @@
 // Copyright (C) 2017-2018 John A. De Goes. All rights reserved.
 package scalaz.zio
-import scalaz.zio.ExitResult.Cause
 
 /**
  * The `Async` class describes the return value of an asynchronous effect
- * that is imported into an `IO` value.
- *
- * Asynchronous effects can return `later`, which represents an uninterruptible
- * asynchronous action, `now` which represents a synchronously computed value,
- * `maybeLater`, which represents an interruptible asynchronous action or `maybeLaterIO`
- * which represents an interruptible asynchronous action where the canceler has the
- * form `Throwable => IO[Nothing, Unit]`
+ * that is imported into an `IO` value. Asynchronous effects can return
+ * values synchronously via returning `now`, or promise to invoke a
+ * callback via returning `later`.
  */
-sealed abstract class Async[+E, +A] extends Product with Serializable { self =>
-  final def fold[E1, B](
-    f: Cause[E] => ExitResult[E1, B],
-    g: A => ExitResult[E1, B]
-  ): Async[E1, B] =
-    self match {
-      case Async.Now(r)            => Async.Now(r.fold(f, g))
-      case x @ Async.MaybeLater(_) => x
-    }
-}
+sealed abstract class Async[+E, +A] extends Product with Serializable
 
 object Async extends Serializable {
-  // TODO: Optimize this common case to less overhead with opaque types
-  final case class Now[E, A](value: ExitResult[E, A]) extends Async[E, A]
-  final case class MaybeLater(canceler: Canceler)     extends Async[Nothing, Nothing]
+  final case class Now[E, A](value: IO[E, A]) extends Async[E, A]
```

Comment:
Since they appear in the API (which means the user can / must use them for `async0`), and there's no harm in exposing them, I think it's better to make them public.

We could hide them with a `fold` method on `Async` that exposes both cases. There's an argument for doing that for performance reasons because then we can avoid allocating the `Now` wrapper.